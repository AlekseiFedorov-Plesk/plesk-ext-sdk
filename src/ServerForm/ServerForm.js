// Copyright 1999-2018. Plesk International GmbH. All rights reserved.

import { createElement, Component, Form, PropTypes, Alert } from '@plesk/ui-library';
import { withRouter } from 'react-router-dom';
import { withApi } from '../Api';
import { withConfig } from '../Config';

class ServerForm extends Component {
    static propTypes = {
        config: PropTypes.shape({
            baseUrl: PropTypes.string.isRequired,
        }).isRequired,
        action: PropTypes.string.isRequired,
        successUrl: PropTypes.string,
        cancelUrl: PropTypes.string,
        children: PropTypes.node,
        history: PropTypes.shape({
            push: PropTypes.func.isRequired,
        }).isRequired,
        api: PropTypes.shape({
            post: PropTypes.func.isRequired,
        }).isRequired,
        onSubmit: PropTypes.func,
        onSuccess: PropTypes.func,
        applyButton: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.object,
        ]),
        cancelButton: PropTypes.oneOfType([
            PropTypes.bool,
            PropTypes.object,
        ]),
        staticContext: PropTypes.any,
    };

    static defaultProps = {
        successUrl: null,
        cancelUrl: null,
        children: null,
        onSubmit: null,
        onSuccess: null,
        applyButton: false,
        cancelButton: false,
        staticContext: null,
    };

    state = {
        state: null,
        errors: {},
    };

    handleCancel = () => {
        const { cancelUrl, history, cancelButton } = this.props;

        if (cancelButton && typeof cancelButton.onClick === 'function' && cancelButton.onClick() === false) {
            return;
        }

        if (cancelUrl) {
            history.push(cancelUrl);
        }
    };

    handleSubmit = values => {
        const { config, action, onSubmit, api } = this.props;

        if (typeof onSubmit === 'function' && onSubmit(values) === false) {
            return;
        }

        this.setState({
            state: 'submit',
        });

        api.post(`${config.baseUrl}${action}`, values).then(({ status, errors, data }) => {
            if (status === 'success') {
                this.setState({
                    state: null,
                    errors,
                }, () => {
                    const { successUrl, history, onSuccess } = this.props;

                    if (typeof onSuccess === 'function' && onSuccess(data) === false) {
                        return;
                    }

                    if (successUrl) {
                        history.push(successUrl);
                    }
                });
            } else {
                this.setState({
                    state: null,
                    errors,
                });
            }
        });
    };

    prepareCancelButton() {
        const { cancelUrl, cancelButton } = this.props;

        if (cancelButton) {
            return {
                ...cancelButton,
                onClick: this.handleCancel,
            };
        }

        if (cancelUrl !== null) {
            return {
                onClick: this.handleCancel,
            };
        }

        return false;
    }

    render() {
        const { state, errors } = this.state;
        const { children, config, successUrl, cancelUrl, history, onSubmit, cancelButton, onSuccess, api, staticContext, ...props } = this.props;
        const formErrors = errors && !Array.isArray(errors) ? errors : null;

        return (
            <Form
                state={state}
                errors={formErrors}
                {...props}
                onSubmit={this.handleSubmit}
                cancelButton={this.prepareCancelButton()}
            >
                {Array.isArray(errors) && errors.map((error, index) => (
                    // eslint-disable-next-line react/no-array-index-key
                    <Alert key={index} intent="danger">
                        {error}
                    </Alert>
                ))}
                {children}
            </Form>
        );
    }
}

export default withRouter(withApi(withConfig(ServerForm)));
