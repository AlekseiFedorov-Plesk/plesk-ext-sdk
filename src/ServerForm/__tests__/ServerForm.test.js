// Copyright 1999-2018. Plesk International GmbH. All rights reserved.

import { createElement, Form } from '@plesk/ui-library';
import { ServerForm } from '../ServerForm';
import { shallow } from 'enzyme';

describe('ServerForm', () => {
    let props;

    beforeEach(() => {
        props = {
            config: {
                baseUrl: '',
            },
            history: {
                push: jest.fn(),
            },
            api: {
                post: jest.fn(),
            },
            statusMessages: { add: jest.fn() },
            action: '/api/save',
        };
    });

    it('renders correctly', () => {
        const wrapper = shallow(
            <ServerForm
                {...props}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('submits data with success', () => {
        const promise = Promise.resolve({ status: 'success' });
        props.api.post.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerForm
                {...props}
                successUrl="/overview"
            />
        );

        wrapper.find(Form).simulate('submit', { username: 'Bob' });
        expect.assertions(2);

        return promise.then(() => {
            expect(props.api.post).toHaveBeenCalledWith('/api/save', { username: 'Bob' });
            expect(props.history.push).toHaveBeenCalledWith('/overview');
        });
    });

    it('cancels form', () => {
        const wrapper = shallow(
            <ServerForm
                {...props}
                cancelUrl="/overview"
            />
        );

        wrapper.find(Form).prop('cancelButton')
            .onClick();
        expect(props.history.push).toHaveBeenCalledWith('/overview');
    });

    it('renders success message correctly', () => {
        const promise = Promise.resolve({ status: 'success' });
        props.api.post.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerForm
                {...props}
                successMessage="Form saved"
            />
        );

        wrapper.find(Form).simulate('submit');
        expect.assertions(1);

        return promise.then(() => {
            expect(props.statusMessages.add).toHaveBeenCalledWith({ intent: 'success', message: 'Form saved' });
        });
    });

    it('renders common error correctly', () => {
        const promise = Promise.resolve({ status: 'error', errors: ['Some error'] });
        props.api.post.mockReturnValueOnce(promise);
        const wrapper = shallow(
            <ServerForm
                {...props}
            />
        );

        wrapper.find(Form).simulate('submit');
        expect.assertions(1);

        return promise.then(() => {
            expect(props.statusMessages.add).toHaveBeenCalledWith({ intent: 'danger', message: 'Some error' });
        });
    });
});
