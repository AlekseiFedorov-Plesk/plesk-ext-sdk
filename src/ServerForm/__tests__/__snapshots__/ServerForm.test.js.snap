// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerForm renders correctly 1`] = `
<Form
  action="/api/save"
  additionalButtons={Array []}
  applyButton={false}
  baseClassName="pul-form"
  cancelButton={false}
  className={null}
  errors={Object {}}
  footerClassName={null}
  hideRequiredLegend={false}
  onSubmit={[Function]}
  requiredMark={
    <span
      className="pul-form__required"
    >
      *
    </span>
  }
  state={null}
  submitButton={true}
  values={Object {}}
  vertical={null}
/>
`;
